Publish-PSResource: Pack and Push Feature

jshigetomi opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Fixes: #1635 
<!-- Summarize your PR between here and the checklist. -->
Decomposed Publish-PSResource to Pack and Push model with an addition of Compress-PSResource command.

Compress-PSResource requires a path to a PS module/script and a destination path to save the resulting nupkg.

Added -Nupkg parameter to Publish-PSResource that pushes the .nupkg file to a specified repository.

Publish-PSResource retains old functionality

## PR Context

This change is to meet an SFI2 requirement to ensure signed packages. This PR enables publishing signed packages through Publish-PSResource by first packing (Compress-PSResource) then signing (User's Choice) then pushing(Publish-PSResource -Nupkg <PathToNupkg>.
<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed]()
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

alerickson reviewed (changes requested):
Looks really good, just made some small suggestions here and there.

jshigetomi commented on src/code/PSResourceHelper.cs:
```diff
@@ -0,0 +1,1214 @@
+            }
+            finally
+            {
+                if(_callerCmdlet == CallerCmdlet.CompressPSResource)
```
The CompressPSResource path deletes it here. But the PublishPSResource path still needs it for push to preserve the previous functionality. The PushResource method finally block handles that.

jshigetomi commented on src/code/CompressPSResource.cs:
```diff
@@ -0,0 +1,71 @@
+    [Cmdlet(VerbsData.Compress,
+        "PSResource",
+        SupportsShouldProcess = true)]
+    [Alias("cmpres")]
```
I didn't know you could do that! I will use for any aliases I have to come up with in the future.

jshigetomi commented on src/code/CompressPSResource.cs:
```diff
@@ -0,0 +1,71 @@
+            // This is to create a better experience for those who have just installed v3 and want to get up and running quickly
+            RepositorySettings.CheckRepositoryStore();
+
+            _psResourceHelper = new PSResourceHelper(
```
Confirm
*Are you sure you want to perform this action?
Performing the operation "Compress-PSResource" on target "CompressPSResource 'C:\Program
Files\PowerShell\7\Modules\PowerShellGet' from the machine".
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"):

more like this maybe?

jshigetomi commented on src/code/PSResourceHelper.cs:
```diff
@@ -0,0 +1,1214 @@
+        internal void PackResource() 
+        {
+            // Returns the name of the file or the name of the directory, depending on path
+            if (!_cmdlet.ShouldProcess(string.Format("{0} '{1}' from the machine", _callerCmdlet, resolvedPath)))
```
Confirm
**Are you sure you want to perform this action?
Performing the operation "Compress-PSResource" on target "CompressPSResource 'C:\Program
Files\PowerShell\7\Modules\PowerShellGet' from the machine".**
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"):

more like this maybe?

jshigetomi commented on src/code/PSResourceHelper.cs:
```diff
@@ -0,0 +1,1214 @@
+                    return;
+                }
+
+                if (_callerCmdlet == CallerCmdlet.CompressPSResource)
```
I will remove this check because it is redundant. Compress-PSResource has no way of changing SkipModuleManifestValidate parameter because did not add it as a parameter.

I think adding an option to check for manifest could be useful to users who want a manifest file. Publish provides the option as well. It would be annoying if the user could only check with publish after they had already packed it.

alerickson replied:
I guess there's a couple things here-- first whether validating the module manifest should happen or not.  I think it should by default always occur, but we can add a parameter to `Compress-PSResource` to skip validation, similar to `Publish-PSResource` (https://github.com/PowerShell/PSResourceGet/blob/d0e71b2cf558822c040143f76803d6c5efb65584/src/code/PublishPSResource.cs#L82).  

Second, if we by default always validate the module manifest then the default value for `SkipModuleManifestValidate` should be false so the module manifest is always checked unless otherwise specified not to (https://github.com/PowerShell/PSResourceGet/blob/5b64c5eb1ccd13074d42d0aaea48c2fa5f6d0f69/src/code/PSResourceHelper.cs#L59)

jshigetomi commented on src/code/PSResourceHelper.cs:
```diff
@@ -0,0 +1,1214 @@
+
+                if (_callerCmdlet == CallerCmdlet.PublishPSResource)
+                {
+                    outputNupkgDir = System.IO.Path.Combine(outputDir, "nupkg");
```
I've added it to their unique constructors for the PublishHelper class.

outputDir is created in PSResource helper so I would have to create it Publish and Compress, then pass it to PackResource(). Then in the Publish path I would have to assign it as a member because pack and push both use it.

jshigetomi commented on src/code/PSResourceHelper.cs:
```diff
@@ -0,0 +1,1214 @@
+                {
+                    if(_isNupkgPathSpecified)
+                    {
+                        outputNupkgDir = pathToNupkgToPublish;
```
passing outputNupkgDir makes it difficult to maintain old PublishResource Behavior. All the outputNupkgDir logic needs to be done in Publish and not in helper and that means in Compress as well.

alerickson commented:
/azp run PowerShell.PSResourceGet

adityapatwardhan commented:
Closing and re-opening to re-trigger CI

adityapatwardhan commented:
/azp run PowerShell.PSResourceGet

alerickson commented:
/azp run PowerShell.PSResourceGet

jshigetomi commented:
Fixed a test that was strangely passing on local machine but failing on git.

The problem was with 
>Get-ChildItem $script:repositoryPath2 | Should -Be $expectedPath
changed to 
>(Get-ChildItem $script:repositoryPath2).FullName | Should -Be $expectedPath

alerickson commented:
/azp run PowerShell.PSResourceGet

alerickson commented on test/PublishPSResourceTests/PublishPSResource.Tests.ps1:
```diff
@@ -567,6 +567,8 @@ Describe "Test Publish-PSResource" -tags 'CI' {
         $expectedPath = Join-Path -Path $script:repositoryPath  -ChildPath "$scriptName.$scriptVersion.nupkg"
         (Get-ChildItem $script:repositoryPath).FullName | Should -Be $expectedPath
+
+        Remove-Item -Path $scriptPath
```
This should probably be moved into the "After All" or "After Each" (whichever is appropriate), just to isolate the test functionality itself and have clean up done outside of test

jshigetomi replied:
This test doesn't delete the test.ps1 file it creates. It only happen for this test.
Maybe it can be like the first test where it removes it in the test?
```pwsh
    It "Publish module with required module not installed on the local machine using -SkipModuleManifestValidate" {
        # Skip the module manifest validation test, which fails from the missing manifest required module.
        $testModulePath = Join-Path -Path $TestDrive -ChildPath ModuleWithMissingRequiredModule
        Publish-PSResource -Path $testModulePath -Repository $testRepository2 -Confirm:$false -SkipDependenciesCheck -SkipModuleManifestValidate

        $expectedPath = Join-Path -Path $script:repositoryPath2  -ChildPath 'ModuleWithMissingRequiredModule.1.0.0.nupkg'
        $publishedModuleFound = Test-Path -Path $expectedPath
        $publishedModuleFound | Should -BeTrue

        if ($publishedModuleFound) {
            Remove-Item $expectedPath -Force -ErrorAction SilentlyContinue
        }
    }
```

alerickson replied:
Oh, sorry I was just looking at the line kind of out of context.  Really the entire line should be removed because `$scriptPath` will be deleted during clean up in the "After Each" block already: https://github.com/PowerShell/PSResourceGet/blob/0b4d77fa43f905284acee01db56f12eea3d3e795/test/PublishPSResourceTests/PublishPSResource.Tests.ps1#L108

alerickson commented on src/code/PublishHelper.cs:
```diff
@@ -0,0 +1,1211 @@
+
+        private bool PackNupkg(string outputDir, string outputNupkgDir, string nuspecFile, out ErrorRecord error)
+        {
+            _cmdlet.WriteDebug($"In {_callerCmdlet}::PackNupkg()");
```
These debug messages should specify "PublishHelper" instead of "_callerCmdlet" to stay consistent with how we're using the debug/verbose messages elsewhere in the repo

jshigetomi replied:
done. Should I also change the errors as well like...

```csharp
                    _cmdlet.WriteError(new ErrorRecord(
                        new ArgumentException($"The resource repository '{Repository}' is not a registered. Please run 'Register-PSResourceRepository' in order to publish to this repository."),
                        "RepositoryNotFound",
                        ErrorCategory.ObjectNotFound,
                        this._cmdlet));

                    return;
```

alerickson replied:
You can just leave it as `this` without the `.cmdlet` (eg: https://github.com/PowerShell/PSResourceGet/blob/e79f6a242f2e90545457b04336b937a47cfcce42/src/code/FindHelper.cs#L110).  And looking at that example, could you change `_cmdlet` to `_cmdletPassedIn`?  Staying consistent with the naming convention just helps us in the future if there's changes we need to make to the entire code base.  It's easier to search and replace one term instead of looking up the different permutations

alerickson commented:
/azp run PowerShell.PSResourceGet

alerickson commented:
/azp run PowerShell.PSResourceGet

alerickson commented:
/azp run PowerShell.PSResourceGet

adityapatwardhan commented:
/azp run PowerShell.PSResourceGet

alerickson commented on src/code/.vs/Microsoft.PowerShell.PSResourceGet/v17/.suo:
Can you add this file back just to keep this PR clean with changes related to Publish/Compress?

jshigetomi replied:
Add the file back.

alerickson commented on test/testFiles/testScripts/test.ps1:
We actually don't need the script as a separate file, the cmdlet `New-PSScriptFileInfo` can create a new script.  The testScripts directory just contains invalid or complicated to create scripts (scenarios that would fail if trying to use `New-PSScriptFileInfo`)

alerickson replied:
The docs for the cmdlet are here: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.psresourceget/new-psscriptfileinfo?view=powershellget-3.x

jshigetomi replied:
This is the test script that is created in that one test where I put remove-item. It was added here on accident when rebasing the PR. I will remove it.

alerickson commented:
/azp run PowerShell.PSResourceGet

adityapatwardhan commented:
/azp run PowerShell.PSResourceGet

adityapatwardhan commented:
/azp run PowerShell.PSResourceGet

adityapatwardhan commented:
/azp run PowerShell.PSResourceGet

alerickson commented:
/azp run PowerShell.PSResourceGet

aluty commented:
@adityapatwardhan Checking back after a while.  I see "preview" has an official release now.

- [PowerShell Gallery | Microsoft.PowerShell.PSResourceGet 1.1.0](https://www.powershellgallery.com/packages/Microsoft.PowerShell.PSResourceGet/1.1.0) (2025-01-09)

Tested and documented somewhere?
